File: Practice2019Offseason/src/main/cpp/F310.cpp
#include "frc/Joystick.h"
#include "frc/buttons/JoystickButton.h"
#include "frc/buttons/POVButton.h"
#include "F310.h"
#include "RobotMap.h"

F310::F310() 
{
    f310joystick = new frc::Joystick(RobotMap::f310);

    greenButtonObject = new frc::JoystickButton(f310joystick, F310::green_button);
    redButtonObject = new frc::JoystickButton(f310joystick, F310::red_button);
    blueButtonObject = new frc::JoystickButton(f310joystick, F310::blue_button);
    orangeButtonObject = new frc::JoystickButton(f310joystick, F310::orange_button);
    leftShoulderButtonObject = new frc::JoystickButton(f310joystick, F310::left_shoulder_button);
    rightShoulderButtonObject = new frc::JoystickButton(f310joystick, F310::right_shoulder_button);
    backButtonObject = new frc::JoystickButton(f310joystick, F310::back_button);
    startButtonObject = new frc::JoystickButton(f310joystick, F310::start_button);
    leftJoyButtonObject = new frc::JoystickButton(f310joystick, F310::left_joy_button);
    rightJoyButtonObject = new frc::JoystickButton(f310joystick, F310::right_joy_button);
    // startButtonObject::whenPressed(new StopAll());
    // blueButtonObject::whenPressed(new ToggleCompressor());

}

double F310::getLeftX() {
    return deadzone(0);
}

double F310::getLeftXCubed() {
    return pow(getLeftX(), 3);
}

double F310::getLeftY() {
    return -deadzone(1); // Negates the Y axis of the joystick so that up is positive and down is
                            // negative
}

double F310::getLeftYCubed() {
    return pow(getLeftY(), 3);
}

double F310::getLeftTrigger() {
    return deadzone(2);
}

double F310::getRightTrigger() {
    return deadzone(3);
}

double F310::getRightX() {
    return deadzone(4);
}

double F310::getRightXCubed() {
    return pow(getRightX(), 3);
}

// ControlWinch
double F310::getRightY() {
    return -deadzone(5); // Negates the Y axis of the joystick so that up is positive and down is
                            // negative
}

bool F310::getButton(int buttonPort) {
    return f310joystick->GetRawButton(buttonPort);
}

bool F310::getButtonPressed(int buttonPort) {
    return f310joystick->GetRawButtonPressed(buttonPort);
}

bool F310::getButtonReleased(int buttonPort) {
    return f310joystick->GetRawButtonReleased(buttonPort);
}

int F310::getDPadPOV() {
    return f310joystick->GetPOV();
}

double F310::deadzone(int port) {
    double joystickValue = f310joystick->GetRawAxis(port);
    double joystickOffset = 0.075;
    double absJoystickValue = abs(joystickValue);
    if (absJoystickValue > joystickOffset) {
        double speed = absJoystickValue;
        speed = (speed * speed) + joystickOffset;
        if (joystickValue > 0)
            return speed;
        else
            return -speed;
    } else {
        return 0;
    }
}

File: TestPrograms2019/TestProgram/src/main/java/frc/robot/F310.java
package frc.robot;

import frc.robot.commands.ToggleCompressor;
import edu.wpi.first.wpilibj.Joystick;
import edu.wpi.first.wpilibj.buttons.Button;
import edu.wpi.first.wpilibj.buttons.JoystickButton;

public class F310 {

	// x on back switch, mode light off
	private Joystick f310;

	public Button greenButtonObject;
	public Button redButtonObject;
	public Button blueButtonObject;
	public Button orangeButtonObject;
	public Button leftShoulderButtonObject;
	public Button rightShoulderButtonObject;
	public Button backButtonObject;
	public Button startButtonObject;
	public Button leftJoyButtonObject;
	public Button rightJoyButtonObject;

	public static final int greenButton = 1;
	public static final int redButton = 2;
	public static final int blueButton = 3;
	public static final int orangeButton = 4;
	public static final int leftShoulderButton = 5;
	public static final int rightShoulderButton = 6;
	public static final int backButton = 7;
	public static final int startButton = 8;
	public static final int leftJoyButton = 9;
	public static final int rightJoyButton = 10;

	public F310() {
		f310 = new Joystick(RobotMap.F310);

		greenButtonObject = new JoystickButton(f310, F310.greenButton);
		redButtonObject = new JoystickButton(f310, F310.redButton);
		blueButtonObject = new JoystickButton(f310, F310.blueButton);
		orangeButtonObject = new JoystickButton(f310, F310.orangeButton);
		leftShoulderButtonObject = new JoystickButton(f310, F310.leftShoulderButton);
		rightShoulderButtonObject = new JoystickButton(f310, F310.rightShoulderButton);
		backButtonObject = new JoystickButton(f310, F310.backButton);
		startButtonObject = new JoystickButton(f310, F310.startButton);
		leftJoyButtonObject = new JoystickButton(f310, F310.leftJoyButton);
		rightJoyButtonObject = new JoystickButton(f310, F310.rightJoyButton);

		blueButtonObject.whenPressed(new ToggleCompressor());

	}

	public double getLeftX() {
		return deadzone(0);
	}

	public double getLeftXSquared() {
		return Math.pow(getLeftX(), 3);
	}

	public double getLeftY() {
		return -deadzone(1); // Negates the Y axis of the joystick so that up is positive and down is
								// negative
	}

	public double getLeftYCubed() {
		return Math.pow(getLeftY(), 3);
	}

	public double getLeftTrigger() {
		return deadzone(2);
	}

	public double getRightTrigger() {
		return deadzone(3);
	}

	public double getRightX() {
		return deadzone(4);
	}

	public double getRightXCubed() {
		return Math.pow(getRightX(), 3);
	}

	// ControlWinch
	public double getRightY() {
		return -deadzone(5); // Negates the Y axis of the joystick so that up is positive and down is
								// negative
	}

	public boolean getButton(int buttonPort) {
		return f310.getRawButton(buttonPort);
	}

	public boolean getButtonPressed(int buttonPort) {
		return f310.getRawButtonPressed(buttonPort);
	}

	public boolean getButtonReleased(int buttonPort) {
		return f310.getRawButtonReleased(buttonPort);
	}

	public int getDPadPOV() {
		return f310.getPOV();
	}

	public double deadzone(int port) {
		double joystickValue = f310.getRawAxis(port);
		double joystickOffset = 0.075;
		double absJoystickValue = Math.abs(joystickValue);
		if (absJoystickValue > joystickOffset) {
			double speed = absJoystickValue;
			speed = (speed * speed) + joystickOffset;
			if (joystickValue > 0)
				return speed;
			else
				return -speed;
		} else {
			return 0;
		}
	}

}
